### PR TITLE
Define properties on Struct.with automatically

### PR DESCRIPTION
**Pull request by jacogr**
Closes https://github.com/polkadot-js/api/issues/419

Sorry for the "revert revert" commit - I pushed this one to master by mistake, then reverted, then branched, then reverted the revert. So messy and fat-fingered.

### CODE REVIEW

File: packages/types/src/codec/Struct.ts [hunk 1]
```diff
@@ -111,6 +111,13 @@ export default class Struct<
     return class extends Struct<S> {
       constructor (value?: any, jsonMap?: Map<keyof S, string>) {
         super(Types, value, jsonMap);
+
+        (Object.keys(Types) as Array<keyof S>).forEach((key, index) => {
+          Object.defineProperty(this, key, {
+            enumerable: true,
+            get: () => this.getAtIndex(index)
```
Comment (amaurymartiny) [inline packages/types/src/codec/Struct.ts]: `this.get(key)` is faster, we don't do an .toArray() each time.

File: packages/types/src/codec/Struct.ts [hunk 2]
```diff
@@ -111,6 +111,13 @@ export default class Struct<
     return class extends Struct<S> {
       constructor (value?: any, jsonMap?: Map<keyof S, string>) {
         super(Types, value, jsonMap);
+
+        (Object.keys(Types) as Array<keyof S>).forEach((key, index) => {
+          Object.defineProperty(this, key, {
```
Comment (amaurymartiny) [inline packages/types/src/codec/Struct.ts]: Does this by any chance add TS typings on the struct?

File: packages/types/src/codec/Struct.ts (hunk 1 above)
Comment (jacogr) [reply packages/types/src/codec/Struct.ts]: Yes, actually had it as `get(key)` originally, will revert. Thanks.

File: packages/types/src/codec/Struct.ts (hunk 2 above)
Comment (jacogr) [reply packages/types/src/codec/Struct.ts]: Not 100% sure what you mean - if it is available, nope, look at the test, had to any to extract it :(

**Pull request merged**